File: cpp-solutions/90.cpp
class Solution {
private:
    set<vector<int>> ss;
    vector<int> vec;
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        dfs(nums,0);
        vector<vector<int>> res;
        for(auto it=ss.begin(); it!=ss.end(); it++){
            res.push_back(*it);
        }
        return res;
    }
    
    void dfs(vector<int>&nums, int cur){
        if(cur==nums.size()){
            ss.insert(vec);
            return;
        }
        dfs(nums, cur+1);
        vec.push_back(nums[cur]);
        dfs(nums, cur+1);
        vec.pop_back();
    }
};

File: golang/90/90.go
package main

import "sort"

var res [][]int
var cur []int

func subsetsWithDup(nums []int) [][]int {
	res = make([][]int, 0)
	cur = make([]int, 0)

	sort.Ints(nums)
	backtrack(nums, 0)
	return res
}

func backtrack(nums []int, start int) {
	tmp := make([]int, len(cur))
	copy(tmp, cur)
	res = append(res, tmp)

	for i := start; i < len(nums); i++ {
		if i > start && nums[i] == nums[i-1] {
			continue
		}
		cur = append(cur, nums[i])
		backtrack(nums, i+1)
		cur = cur[:len(cur)-1]
	}

	return
}
